### PR TITLE
Unrelated expired trusted CA certificate should not throw an exception

### PR DESCRIPTION
**Pull request by rabadashTheFool**
WE2-913

Signed-off-by: Mihkel Kivisild mihkel.kivisild@cgi.com

### CODE REVIEW

**Review by metsma — APPROVED** [review]

File: src/WebEid.Security/Util/X509CertificateExtensions.cs [hunk 1]
```diff
@@ -80,6 +82,7 @@ public static X509Certificate2 ValidateIsSignedByTrustedCa(this X509Certificate2
                     UrlRetrievalTimeout = TimeSpan.Zero
                 }
             };
+
```
Comment (mrts) [inline src/WebEid.Security/Util/X509CertificateExtensions.cs:85]: I don't see where you check CA certificate expiry? I expected to see it below around line 117, something like this:

```c#
ValidateCertificateExpiry(chainElement.Certificate, DateTimeProvider.UtcNow, "Trusted CA");
return chainElement.Certificate;
```

Or is this somehow handled implicitly?

File: src/WebEid.Security/Validator/CertValidators/SubjectCertificateTrustedValidator.cs [hunk 2]
```diff
@@ -39,14 +40,18 @@ public SubjectCertificateTrustedValidator(ICollection<X509Certificate2> trustedC
         }
 
         /// <summary>
-        /// Validates that the user certificate from the authentication token is signed by a trusted certificate authority.
+        /// Checks that the user certificate from the authentication token is valid and signed by
+        /// a trusted certificate authority.Also checks the validity of the user certificate's
```
Comment (mrts) [inline src/WebEid.Security/Validator/CertValidators/SubjectCertificateTrustedValidator.cs:44]: Minor missing space:
```suggestion
        /// a trusted certificate authority. Also checks the validity of the user certificate's
```

File: src/WebEid.Security/Util/X509CertificateExtensions.cs [hunk 3]
```diff
@@ -69,6 +69,8 @@ public static void ValidateCertificateExpiry(this Org.BouncyCastle.X509.X509Cert
 
         public static X509Certificate2 ValidateIsSignedByTrustedCa(this X509Certificate2 certificate, ICollection<X509Certificate2> trustedCaCertificates)
```
Comment (mrts) [inline src/WebEid.Security/Util/X509CertificateExtensions.cs:70]: I would like to make the method name intention revealing:
```suggestion
        public static X509Certificate2 ValidateIsValidAndSignedByTrustedCa(this X509Certificate2 certificate, ICollection<X509Certificate2> trustedCaCertificates)
```
I know this is missing in Java, but I plan to change it.

**Review by mrts — CHANGES_REQUESTED** [review]

File: src/WebEid.Security/Util/X509CertificateExtensions.cs (hunk 1 above)
Comment (rabadashTheFool) [reply src/WebEid.Security/Util/X509CertificateExtensions.cs:85]: Good catch. I will fix this promptly.

File: src/WebEid.Security/Validator/CertValidators/SubjectCertificateTrustedValidator.cs (hunk 2 above)
Comment (rabadashTheFool) [reply src/WebEid.Security/Validator/CertValidators/SubjectCertificateTrustedValidator.cs]: Fixed with latest commit

File: src/WebEid.Security/Util/X509CertificateExtensions.cs (hunk 3 above)
Comment (rabadashTheFool) [reply src/WebEid.Security/Util/X509CertificateExtensions.cs]: I renamed the method with the latest commit.

**Comment by rabadashTheFool** [comment]
While working on this and OCSP time fix, I discovered that due to the `DateTimeProvider.overridenUtcNow` being `ThreadStatic`, some test were not working as intended.
Some tests were seemingly working while in reality they should not. This is mainly due to the `SubjectCertificateNotRevokedValidator` validation being done asynchronously. This created a new `DateTimeProvider` instance which was different from the one created at the test.
Now the tests should work as intented and should be working like they are in the Java library.

**Review by mrts — APPROVED** [review]

**Pull request merged**